modules/fdb/examples/sqlite-to-fdb.rs

```diff
@@ -0,0 +1,155 @@
+        .wrap_err_with(|| format!("Failed to open fdb template '{}'", opts.template.display()))?;
+    let mmap = unsafe { Mmap::map(&template_file)? };
+    let buffer: &[u8] = &mmap;
+    let template_db = Database::new(buffer);
+    
+    // sqlite input
+    let conn = Connection::open_with_flags(&opts.src, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)
+        .wrap_err_with(|| format!("Failed to open sqlite file '{}'", opts.src.display()))?;
+
+    // fdb output
+    let dest_file = File::create(&opts.dest)
+        .wrap_err_with(|| format!("Failed to crate output file '{}'", opts.dest.display()))?;
+    let mut dest_out = BufWriter::new(dest_file);
+    let mut dest_db = store::Database::new();
+
+    for template_table in template_db.tables()?.iter() {
+        let template_table = template_table?;
+
+        // Find number of unique values in first column of source table
+        let unique_key_count = conn.query_row::<u32, _, _>(
+            &format!(
+                "select count(distinct [{}]) as unique_count from {}",
+                template_table.column_at(0).unwrap().name(),
+                template_table.name()
+            ),
+            rusqlite::NO_PARAMS,
+            |row| row.get(0),
+        )?;
+
+        // Bucket count should be 0 or a power of two
+        let new_bucket_count = if unique_key_count == 0 {
+            0
+        } else {
+            u32::next_power_of_two(unique_key_count)
+        };
+
+        // Create destination table
+        let mut dest_table = store::Table::new(new_bucket_count as usize);
+
+        // Number of columns destination table should have
+        let column_count = template_table.column_count();
+
+        // Vector to store target datatypes in as these can't be determined from the sqlite source db
+        let mut target_types: Vec<ValueType> = Vec::with_capacity(column_count);
+
+        // Build select query with the same column names and order as the template FDB
+        let mut select_query = String::from("select ");
+
+        // Write select query and store target datatypes
+        for index in 0..column_count {
+            let template_column = template_table
+                .column_at(index).unwrap();
+
+            let template_column_name = template_column.name();
+
+            write!(select_query, "[{}]", template_column_name).unwrap();
+
+            if index < column_count - 1 {
+                write!(select_query, ", ").unwrap();
+            }
+
+            target_types.push(template_column.value_type());
+
+            dest_table.push_column(template_column.name_raw(), template_column.value_type());
+        }
+
+        select_query.push_str(" from ");
+        select_query.push_str(&template_table.name());
+
+        // Execute query
+        let mut statement = conn.prepare(&select_query)?;
+        let mut rows = statement.query(rusqlite::NO_PARAMS)?;
+
+        // Iterate over rows
+        while let Some(sqlite_row) = rows.next()? {
```

Comment:
Is there a reason `for sqlite_row in rows` isn't used here (with `let sqlite_row = sqlite_row?;`) ?

Reply:
rusqlite's `Rows` does not implement the `Iterator` trait, so it's not possible to use `for ... in` here. It instead implements FallibleStreamingIterator and suggests `while let Some(...) = rows.next()?` as alternative to `for ... in`: https://docs.rs/rusqlite/0.25.3/rusqlite/struct.Rows.html#impl-FallibleStreamingIterator